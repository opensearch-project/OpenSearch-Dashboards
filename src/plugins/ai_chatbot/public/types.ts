/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 */

import { UiActionsStart } from '../../ui_actions/public';

export interface AIChatbotSetupDependencies {
  contextProvider?: any;
}

export interface AIChatbotStartDependencies {
  uiActions: UiActionsStart;
  contextProvider?: any;
}

export interface AIChatbotSetup {}

export interface AIChatbotStart {
  openChatbot: () => void;
  closeChatbot: () => void;
}

export interface ContextData {
  pageType?: string;
  currentUrl?: string;
  panels?: any[];
  expandedDocuments?: any[];
  filters?: any[];
  query?: any;
  timeRange?: any;
  timestamp?: number;
}

export interface AgentTool {
  name: string;
  description: string;
  parameters: {
    type: string;
    properties: Record<string, any>;
    required?: string[];
  };
  execute: (params: any) => Promise<string>;
}