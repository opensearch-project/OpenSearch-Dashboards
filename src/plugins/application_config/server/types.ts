/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { IScopedClusterClient, Headers } from 'src/core/server';

export interface ApplicationConfigPluginSetup {
  getConfigurationClient: (inputOpenSearchClient: IScopedClusterClient) => ConfigurationClient;
  registerConfigurationClient: (inputConfigurationClient: ConfigurationClient) => void;
}
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface ApplicationConfigPluginStart {}

export interface ConfigurationClientOptions {
  headers: Headers;
}

/**
 * The interface defines the operations against the application configurations at both entity level and whole level.
 *
 */
export interface ConfigurationClient {
  /**
   * Get all the configurations.
   *
   * @param {options} options, an optional parameter
   * @returns {Map<string, string>} all the configurations
   */
  getConfig(options?: ConfigurationClientOptions): Promise<Map<string, string>>;

  /**
   * Get the value for the input entity.
   *
   * @param {entity} name of the entity
   * @param {options} options, an optional parameter
   * @returns {string} value of the entity
   */
  getEntityConfig(entity: string, options?: ConfigurationClientOptions): Promise<string>;

  /**
   * Update the input entity with a new value.
   *
   * @param {entity} name of the entity
   * @param {newValue} new configuration value of the entity
   * @param {options} options, an optional parameter
   * @returns {string} updated configuration value of the entity
   */
  updateEntityConfig(
    entity: string,
    newValue: string,
    options?: ConfigurationClientOptions
  ): Promise<string>;

  /**
   * Delete the input entity from configurations.
   *
   * @param {entity} name of the entity
   * @param {options} options, an optional parameter
   * @returns {string} name of the deleted entity
   */
  deleteEntityConfig(entity: string, options?: ConfigurationClientOptions): Promise<string>;
}
