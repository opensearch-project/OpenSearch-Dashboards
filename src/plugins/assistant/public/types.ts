/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { NavigationPublicPluginStart } from '../../navigation/public';
import { ClientConfig } from '../common/types';
import { ContextProviderStart } from '../../context_provider/public';
import { ContextService } from './services/context_service';
import { ActionExecutor } from './services/action_executor';

export interface AssistantPluginSetup {
  getGreeting: () => string;
  getConfig: () => ClientConfig;
  isEnabled: () => boolean;
}

export interface AssistantPluginStart {
  getConfig: () => ClientConfig;
  isEnabled: () => boolean;
  getContextService: () => ContextService;
  getActionExecutor: () => ActionExecutor;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface AssistantPluginSetupDependencies {}

export interface AppPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
  contextProvider: ContextProviderStart;
}
