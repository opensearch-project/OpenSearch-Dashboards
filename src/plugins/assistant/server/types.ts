/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ClientConfig } from '../common/types';
import { AgentClient } from './services';

export interface AssistantPluginSetup {
  getConfig(): ClientConfig;
  getAgentClient(): AgentClient | undefined;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface AssistantPluginStart {}
