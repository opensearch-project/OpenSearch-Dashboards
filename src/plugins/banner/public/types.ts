/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { BannerService } from './services/banner_service';

/**
 * Setup contract for the banner plugin.
 * Intentionally empty for now as the plugin doesn't expose any setup functionality.
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface BannerPluginSetup {}

/**
 * Start contract for the banner plugin.
 */
export interface BannerPluginStart {
  bannerService: BannerService;
}
