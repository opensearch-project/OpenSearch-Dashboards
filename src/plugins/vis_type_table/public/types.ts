/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { SchemaConfig } from 'src/plugins/visualizations/public';
import { IFieldFormat } from 'src/plugins/data/public';

export enum AggTypes {
  SUM = 'sum',
  AVG = 'avg',
  MIN = 'min',
  MAX = 'max',
  COUNT = 'count',
}

export interface TableVisConfig extends TableVisParams {
  title: string;
  metrics: SchemaConfig[];
  buckets: SchemaConfig[];
  splitRow?: SchemaConfig[];
  splitColumn?: SchemaConfig[];
}

export interface TableVisParams {
  perPage: number | '';
  showPartialRows: boolean;
  showMetricsAtAllLevels: boolean;
  showTotal: boolean;
  totalFunc: AggTypes;
  percentageCol: string;
}

export interface FormattedColumn {
  id: string;
  title: string;
  formatter: IFieldFormat;
  filterable: boolean;
  formattedTotal?: string | number;
  sumTotal?: number;
  total?: number;
}

export interface ColumnWidth {
  colIndex: number;
  width: number;
}

export interface ColumnSort {
  colIndex?: number;
  direction?: 'asc' | 'desc';
}
