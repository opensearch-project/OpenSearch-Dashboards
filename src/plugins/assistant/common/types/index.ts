/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export * from './chat';
export * from './agent';
export * from './config';
