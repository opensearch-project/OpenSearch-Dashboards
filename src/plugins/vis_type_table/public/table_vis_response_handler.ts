/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { getFormatService } from './services';
import { OpenSearchDashboardsDatatable } from '../../expressions/public';
import { FormattedColumn, TableVisConfig } from './types';
import { convertToFormattedData } from './utils/convert_to_formatted_data';

export interface Table {
  columns: OpenSearchDashboardsDatatable['columns'];
  rows: OpenSearchDashboardsDatatable['rows'];
}

export interface FormattedTableContext extends Table {
  formattedColumns: FormattedColumn[];
}

export interface TableGroup {
  table: FormattedTableContext;
  title: string;
}

export interface TableVisData {
  table?: FormattedTableContext;
  tableGroups: TableGroup[];
  direction?: 'row' | 'column';
}

export function tableVisResponseHandler(
  input: OpenSearchDashboardsDatatable,
  config: TableVisConfig
): TableVisData {
  let table: FormattedTableContext | undefined;
  const tableGroups: TableGroup[] = [];
  let direction: TableVisData['direction'];

  const split = config.splitColumn || config.splitRow;

  if (split) {
    direction = config.splitRow ? 'row' : 'column';
    const splitColumnIndex = split[0].accessor;
    const splitColumnFormatter = getFormatService().deserialize(split[0].format);
    const splitColumn = input.columns[splitColumnIndex];
    const splitMap: { [key: string]: number } = {};
    let splitIndex = 0;

    input.rows.forEach((row, rowIndex) => {
      const splitValue: any = row[splitColumn.id];

      if (!splitMap.hasOwnProperty(splitValue as any)) {
        (splitMap as any)[splitValue] = splitIndex++;
        const tableGroup: TableGroup = {
          title: `${splitColumnFormatter.convert(splitValue)}: ${splitColumn.name}`,
          table: {
            formattedColumns: [],
            rows: [],
            columns: input.columns,
          },
        };

        tableGroups.push(tableGroup);
      }

      const tableIndex = splitMap[splitValue];
      tableGroups[tableIndex].table.rows.push(row);
    });

    // format tables
    tableGroups.forEach((tableGroup) => {
      tableGroup.table = convertToFormattedData(tableGroup.table, config);
    });
  } else {
    table = convertToFormattedData(
      {
        columns: input.columns,
        rows: input.rows,
      },
      config
    );
  }

  return {
    table,
    tableGroups,
    direction,
  };
}
