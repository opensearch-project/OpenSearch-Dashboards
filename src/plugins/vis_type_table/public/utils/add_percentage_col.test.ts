/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { addPercentageCol } from './add_percentage_col';
import { FormattedColumn } from '../types';
import { Table } from '../table_vis_response_handler';

const mockDeserialize = jest.fn(() => ({}));
jest.mock('../services', () => ({
  getFormatService: jest.fn(() => ({
    deserialize: mockDeserialize,
  })),
}));

let formattedColumns: FormattedColumn[];
const rows = [
  { 'col-0-2': 'Alice', 'col-1-1': 3 },
  { 'col-0-2': 'Anthony', 'col-1-1': 1 },
  { 'col-0-2': 'Timmy', 'col-1-1': 1 },
] as Table['rows'];

beforeEach(() => {
  formattedColumns = [
    {
      id: 'col-0-2',
      title: 'name.keyword: Descending',
      formatter: {},
      filterable: true,
    },
    {
      id: 'col-1-1',
      title: 'Count',
      formatter: {},
      filterable: false,
      sumTotal: 5,
    },
  ] as FormattedColumn[];
});

describe('addPercentageCol', () => {
  it('should add new percentage column', () => {
    const result = addPercentageCol(formattedColumns, 'count', rows, 1);
    expect(result).toMatchSnapshot();
  });

  it('should handle sumTotal being 0', () => {
    formattedColumns[1].sumTotal = 0;
    const result = addPercentageCol(formattedColumns, 'count', rows, 1);
    expect(result).toMatchSnapshot();
  });

  it('should handle empty input data', () => {
    const emptyFormattedColumns: FormattedColumn[] = [];
    const emptyRows: Table['rows'] = [];
    const result = addPercentageCol(emptyFormattedColumns, 'count', emptyRows, 1);
    expect(result).toMatchSnapshot();
  });

  it('should handle input data with one row', () => {
    const oneRow = [{ 'col-0-2': 'Alice', 'col-1-1': 3 }] as Table['rows'];
    const result = addPercentageCol(formattedColumns, 'count', oneRow, 1);
    expect(result).toMatchSnapshot();
  });

  it('should handle input data with null values', () => {
    const nullValueRows = [
      { 'col-0-2': 'Alice', 'col-1-1': null },
      { 'col-0-2': 'Anthony', 'col-1-1': null },
      { 'col-0-2': 'Timmy', 'col-1-1': null },
    ] as Table['rows'];
    const result = addPercentageCol(formattedColumns, 'count', nullValueRows, 1);
    expect(result).toMatchSnapshot();
  });
});
