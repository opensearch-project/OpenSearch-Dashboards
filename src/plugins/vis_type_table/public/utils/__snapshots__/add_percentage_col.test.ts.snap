// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addPercentageCol should add new percentage column 1`] = `
Object {
  "cols": Array [
    Object {
      "filterable": true,
      "formatter": Object {},
      "id": "col-0-2",
      "title": "name.keyword: Descending",
    },
    Object {
      "filterable": false,
      "formatter": Object {},
      "id": "col-1-1",
      "sumTotal": 5,
      "title": "Count",
    },
    Object {
      "filterable": false,
      "formatter": Object {},
      "id": "col-1-1-percents",
      "title": "count percentages",
    },
  ],
  "rows": Array [
    Object {
      "col-0-2": "Alice",
      "col-1-1": 3,
      "col-1-1-percents": 0.6,
    },
    Object {
      "col-0-2": "Anthony",
      "col-1-1": 1,
      "col-1-1-percents": 0.2,
    },
    Object {
      "col-0-2": "Timmy",
      "col-1-1": 1,
      "col-1-1-percents": 0.2,
    },
  ],
}
`;

exports[`addPercentageCol should handle empty input data 1`] = `
Object {
  "cols": Array [],
  "rows": Array [],
}
`;

exports[`addPercentageCol should handle input data with null values 1`] = `
Object {
  "cols": Array [
    Object {
      "filterable": true,
      "formatter": Object {},
      "id": "col-0-2",
      "title": "name.keyword: Descending",
    },
    Object {
      "filterable": false,
      "formatter": Object {},
      "id": "col-1-1",
      "sumTotal": 5,
      "title": "Count",
    },
    Object {
      "filterable": false,
      "formatter": Object {},
      "id": "col-1-1-percents",
      "title": "count percentages",
    },
  ],
  "rows": Array [
    Object {
      "col-0-2": "Alice",
      "col-1-1": null,
      "col-1-1-percents": 0,
    },
    Object {
      "col-0-2": "Anthony",
      "col-1-1": null,
      "col-1-1-percents": 0,
    },
    Object {
      "col-0-2": "Timmy",
      "col-1-1": null,
      "col-1-1-percents": 0,
    },
  ],
}
`;

exports[`addPercentageCol should handle input data with one row 1`] = `
Object {
  "cols": Array [
    Object {
      "filterable": true,
      "formatter": Object {},
      "id": "col-0-2",
      "title": "name.keyword: Descending",
    },
    Object {
      "filterable": false,
      "formatter": Object {},
      "id": "col-1-1",
      "sumTotal": 5,
      "title": "Count",
    },
    Object {
      "filterable": false,
      "formatter": Object {},
      "id": "col-1-1-percents",
      "title": "count percentages",
    },
  ],
  "rows": Array [
    Object {
      "col-0-2": "Alice",
      "col-1-1": 3,
      "col-1-1-percents": 0.6,
    },
  ],
}
`;

exports[`addPercentageCol should handle sumTotal being 0 1`] = `
Object {
  "cols": Array [
    Object {
      "filterable": true,
      "formatter": Object {},
      "id": "col-0-2",
      "title": "name.keyword: Descending",
    },
    Object {
      "filterable": false,
      "formatter": Object {},
      "id": "col-1-1",
      "sumTotal": 0,
      "title": "Count",
    },
    Object {
      "filterable": false,
      "formatter": Object {},
      "id": "col-1-1-percents",
      "title": "count percentages",
    },
  ],
  "rows": Array [
    Object {
      "col-0-2": "Alice",
      "col-1-1": 3,
      "col-1-1-percents": 0,
    },
    Object {
      "col-0-2": "Anthony",
      "col-1-1": 1,
      "col-1-1-percents": 0,
    },
    Object {
      "col-0-2": "Timmy",
      "col-1-1": 1,
      "col-1-1-percents": 0,
    },
  ],
}
`;
